### PR TITLE
fix: add UnionLike type for FieldErrors and FieldNamesMarkedBoolean type

### LINKED ISSUE #6360: [Bug]: Error typings incorrect, when form type is a union type
### Version Number

7.13.0

### Codesandbox/Expo snack

https://codesandbox.io/s/happy-hamilton-hstoc?file=/src/App.tsx

### Steps to reproduce

1. Use a union type with mutually exclusive fields for `useForm`
2. Try to access one of the exclusive fields in `formState.errors`
3. TypeScript will report that no such field exists

### Expected behaviour

No typescript errors for fields defined on the union type.

This is especially relevant, since `zod` is one of the supported validation libraries and it creates union types like that.

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kotarella1110**
## `UnionLike` type examples

```ts
type T1 = UnionLike<
  | {
      kind: 'string';
      stringInput: string;
    }
  | {
      kind: 'number';
      numberInput: number;
    }
>;
// type T1 = {
//     kind: "string" | "number";
//     stringInput?: string | undefined;
//     numberInput?: number | undefined;
// }
type T2 = UnionLike<
  | {
      foo: number;
      bar: string;
    }
  | {
      foo: boolean;
      hoge: {
        fuga: string;
      };
    }
>;
// type T2 = {
//     foo: number | boolean;
//     bar?: string | undefined;
//     hoge?: {
//         fuga: string;
//     } | undefined;
// }
type T3 = UnionLike<{
  nested:
    | {
        foo: string;
        hoge:
          | {
              fuga: string;
            }
          | {
              fuga: boolean;
              fooBar: boolean;
            };
      }
    | {
        bar: number;
      };
}>;
// type T3 = {
//   nested: {
//       foo?: string | undefined;
//       hoge?: {
//           fuga: string | boolean;
//           fooBar?: boolean | undefined;
//       } | undefined;
//       bar?: number | undefined;
//   };
// }
```

## Resolved Ref

- [CodeSandbox](https://codesandbox.io/s/6360-dw4ic?file=/src/App.tsx)
- [TypeScript Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gEgHIkxGADUoIBXEALnnJQEsBHS+NATwFsAjJCAbgBQAkAA8ADkhhYMbMQnyESZSgB4B8eABVSFBKIwgUwNLRQBrFEgDuKANoBdeAB94AJXCTgKzDHooA5gA0phbWKAB8zvAAslBiKnShNsGJljaRALzqGlGpYQ7ZGi7uyDBePn5BIWkRgQKZ8ADe2bZ4BAZKuvY0KCAAbiAwggC+8ABkWjqUggIycvAAqgyoANIgbGgqmg2a8PqGxvBQKGzwAPzwZutIAGZa8D39gzNzCEv0qFMEW8ErO3siAxGEzHNjZC4rAFAw5XNi3LSFc5aWwreyIx4DGDZDHPISveAAeTEGA+KDIaw2W3++2BRxO4Ka8BR8D8l2ud009gAtGcaI1RjTDgAFehgMw-eB-JGQnEweDDWyw+Gc7HwXqYl6yBBC2AksgAITYEshgpMSo5DQJHHoGGNkQmOqk9DIKhFYrt4U183eqAAMvQrlT4BkmZyoQcTLYACJQAxRABi9AgIH9mATSYQLgUHS+aI0F00qtsYdNTPcUGAqAgbAAgjAYFAjaDwnmkY0mZDWeatN1FssUP7A5oUfZIsMiyXARGmUguAArcAYVsXR16iCGtQ5HLNLdb5msn0oCmbba9w+DkAqQ9fE+-cKexEaYZ1XfwACiIkg5FAV-7x6pwTdtsUREiSqDkusJ7hIi0EaDQmheggUYgCAYgfmIxxeMBIa7KWs4LmAGAMrhU60n4NyDIsiIXO2+4oGycJ3AsvbIah6GYVeI5juiapPFicFaIhWgAIzBvArFoeIHHZOeAaXoiLg7q+lx+MANAAOQVAE6mCMprAYL4AQAJIoGI5AYDQWn+Lpu7jluimPjkZiqRpKDkNwgw6Y5GhuR5MAmWZFlqu5XC4q+dnwNBD4APTRfA+KaKJIbNLFu7OUYNAAERWZlUSZb5oUwJlgipVuVkBeZvL6YZ-i5EYIA3H4RAlXFW4FYMFUYFV7Vyi4dCgI1vTAC18qzFqWgAExiRJ7FGJufakhe81FE03nwDcSBID0IVhcpXCwJZBmVDZW4RStSmvhtW3wDwvAgMcJ27gAFkg-jUKtekaDc5D+FAh01Y9p2A6NGhRSNCVTclAilTkV3bX5US3cmD3Q61OT7TAVVWXVA1NcNqPPa9IBVSlaOXT9f3VcdBOnTjDV4yN44JQAzNNKGSRhc0yf2S0XbxihUI5Dmfetm3-dTIsvW9gsi+da16d9v3iwEwN6Wdn3C7LW6K5TSP3SgqufVd+oHTdm3Iwb8u2Yb6tyyLGPw4VNs2cMYMEyzYmkzkvQC6tMPa5tWNHQEdODc1NOE29JMRwrFPK7VLh6yj-vKcbsBVUn9F9fVYf4ynT6hwzMfo+njuUdnuNDSNT6M0AA)

## Resolve Issues

resolve #6360

### CODE REVIEW

**Comment by bluebill1049** [comment]
looks like this issue https://github.com/react-hook-form/react-hook-form/issues/6390 is associated with `DeepMap` as well.

**Comment by bluebill1049** [comment]
wow, pass built! 🥳

File: src/types/utils.ts [hunk 1]
```diff
@@ -10,6 +10,8 @@ export type Primitive =
   | symbol
   | bigint;
 
+export type Builtin = NestedValue | Date | FileList | File;
```
Comment (bluebill1049) [inline src/types/utils.ts:13]: I see, we threat those build-in and native objects separately.

**Comment by bluebill1049** [comment]
`v7.14.3-beta.0`

**Comment by kotarella1110** [comment]
This seems to be working correctly in `v7.14.3-beta.0`.
https://codesandbox.io/s/6360-dw4ic?file=/src/App.tsx

File: src/__tests__/useForm/register.test.tsx [hunk 2]
```diff
@@ -1188,7 +1188,7 @@ describe('register', () => {
 
       const App = () => {
         const { register, handleSubmit } = useForm<{
-          test: Date;
+          test: Date | string;
```
Comment (kotarella1110) [inline src/__tests__/useForm/register.test.tsx:1191]: @bluebill1049 With this PR fix, when users use the `valueAs` option, they must now define it as a Union type.

- valueAsDate: `Date | string`
- valueAsNumber: `number | string`
- setValueAs: `xxx | string`

This is a by-product of fixing the type correctly.
I think you and I discussed earlier whether the `FormValues` type is the field values type or the submit values type. I believe it is both.

Comment (bluebill1049) [reply src/__tests__/useForm/register.test.tsx:1191]: That's expected and correct typing. 👍

File: src/types/utils.ts (hunk 1 above)
Comment (kotarella1110) [reply src/types/utils.ts]: I will do this in another PR.

**Comment by kotarella1110** [comment]
@bluebill1049 Type error will caused if the `FormValues` type is as follows.

```ts
type FormValues = {
  foo: string | {
    bar: string
  }
}

const { formState } = useForm<FormValues>();
// formState.errors: {
//     foo?: FieldError | {
//        bar?: FieldError | undefined;
//    } | undefined;
// }

formState.errors.foo?.bar // type error
```

Should this be fixed?
If we fix this, the `errors` type would be as follows.

```ts
// formState.errors: {
//     foo?: FieldError & {
//        bar?: FieldError | undefined;
//    } | undefined;
// }
```

**Comment by bluebill1049** [comment]
```tsx
type FormValues = {
  foo: string | {
    bar: string
  }
}
```

I think that's a pretty extreme case, when an input transforms from primitive to a nested object, maybe we can leave this one for now unless you have an idea on how to easily support that (just don't want to put too much pressure on you).  The improvement that you have introduced is already a massive milestone for us to support union type. 💪

**Comment by kotarella1110** [comment]
@bluebill1049 please review 🙏🏻

File: src/types/utils.ts [hunk 3]
```diff
@@ -123,3 +123,30 @@ export type FieldPathValues<
...
   >;
 };
+
+type UnionKeys<T> = T extends any ? keyof T : never;
+
+type UnionValues<T, K> = T extends any
+  ? K extends keyof T
+    ? T[K]
+    : never
+  : never;
```
Comment (bluebill1049) [inline src/types/utils.ts:133]: good use of distributive conditional types, learn something new every day.

File: src/types/utils.ts [hunk 4]
```diff
@@ -123,3 +123,30 @@ export type FieldPathValues<
...
+
+type OptionalKeys<T> = T extends any
+  ? { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T]
```
Comment (bluebill1049) [inline src/types/utils.ts:136]: a bit lost in here, getting all the `keys` as required and 
`{} extends Pick<T, K> ? K : never }[keyof T]` this part means?

File: src/types/utils.ts [hunk 5]
```diff
@@ -123,3 +123,30 @@ export type FieldPathValues<
...
+  : [T] extends [ReadonlyArray<any>]
+  ? { [K in keyof T]: UnionLike<T[K]> }
+  : [T] extends [object]
+  ? PartialBy<
+      {
+        [K in UnionKeys<T>]: UnionLike<UnionValues<T, K>>;
+      },
+      Exclude<UnionKeys<T>, keyof T> | OptionalKeys<T>
+    >
+  : T;
```
Comment (bluebill1049) [inline src/types/utils.ts:152]: mind-blowing 👀

File: src/useFormState.ts [hunk 6]
```diff
@@ -38,7 +38,7 @@ function useFormState<TFieldValues extends FieldValues = FieldValues>(
         shouldRenderFormState(formState, _localProxyFormState.current) &&
         updateFormState({
           ...control._formState,
-          ...formState,
+          ...(formState as TFieldValues),
```
Comment (bluebill1049) [inline src/useFormState.ts:41]: curious why we need the cast in here.

**Review by bluebill1049 — DISMISSED** [review]
Looks amazing! ✨ a couple of questions.

File: src/types/utils.ts (hunk 3 above)
Comment (kotarella1110) [reply src/types/utils.ts:133]: 🙌

File: src/types/utils.ts (hunk 4 above)
Comment (kotarella1110) [reply src/types/utils.ts:136]: Please see below.

```ts
type T1 = {} extends { optional?: string } ? true : false; // true
type T2 = {} extends { required: string } ? true : false; // false
```

`{} extends Pick<T, K> ? K : never }[keyof T]` can only get optional keys!

File: src/useFormState.ts (hunk 6 above)
Comment (kotarella1110) [reply src/useFormState.ts]: The cast was not needed and has been fixed: https://github.com/react-hook-form/react-hook-form/pull/6381/commits/a5bf94242caac3cb4984f82762407d2d46a8d1ce

**Comment by kotarella1110** [comment]
@bluebill1049 please re-review 🙌

File: src/types/utils.ts (hunk 4 above)
Comment (bluebill1049) [reply src/types/utils.ts:136]: Thanks!

File: src/useFormState.ts (hunk 6 above)
Comment (bluebill1049) [reply src/useFormState.ts]: sweet! thanks

**Review by bluebill1049 — APPROVED** [review]
LGTM! 🚀

**Pull request merged**